src/graph_gui.cpp

```diff
@@ -9,156 +9,183 @@
+
 	GraphLegendWindow(WindowDesc &desc, WindowNumber window_number) : Window(desc)
 	{
-		this->InitNested(window_number);
-
-		for (CompanyID c = COMPANY_FIRST; c < MAX_COMPANIES; c++) {
-			if (!HasBit(_legend_excluded_companies, c)) this->LowerWidget(WID_GL_FIRST_COMPANY + c);
-
-			this->OnInvalidateData(c);
+		this->CreateNestedTree();
+		if (window_number == 0) {
+			this->selector = &_company_selector;
+		} else {
+			this->selector = &_cargo_selector;
 		}
+		this->selector->Init(this);
+		this->FinishInitNested(window_number);
 	}
 
 	void DrawWidget(const Rect &r, WidgetID widget) const override
 	{
-		if (!IsInsideMM(widget, WID_GL_FIRST_COMPANY, WID_GL_FIRST_COMPANY + MAX_COMPANIES)) return;
-
-		CompanyID cid = (CompanyID)(widget - WID_GL_FIRST_COMPANY);
-
-		if (!Company::IsValidID(cid)) return;
-
-		bool rtl = _current_text_dir == TD_RTL;
-
-		const Rect ir = r.Shrink(WidgetDimensions::scaled.framerect);
-		Dimension d = GetSpriteSize(SPR_COMPANY_ICON);
-		DrawCompanyIcon(cid, rtl ? ir.right - d.width : ir.left, CenterBounds(ir.top, ir.bottom, d.height));
-
-		const Rect tr = ir.Indent(d.width + WidgetDimensions::scaled.hsep_normal, rtl);
-		SetDParam(0, cid);
-		SetDParam(1, cid);
-		DrawString(tr.left, tr.right, CenterBounds(tr.top, tr.bottom, GetCharacterHeight(FS_NORMAL)), STR_COMPANY_NAME_COMPANY_NUM, HasBit(_legend_excluded_companies, cid) ? TC_BLACK : TC_WHITE);
+		this->selector->DrawWidget(r, widget);
 	}
 
-	void OnClick([[maybe_unused]] Point pt, WidgetID widget, [[maybe_unused]] int click_count) override
+	void OnClick(Point pt, WidgetID widget, int click_count) override
 	{
-		if (!IsInsideMM(widget, WID_GL_FIRST_COMPANY, WID_GL_FIRST_COMPANY + MAX_COMPANIES)) return;
-
-		ToggleBit(_legend_excluded_companies, widget - WID_GL_FIRST_COMPANY);
-		this->ToggleWidgetLoweredState(widget);
+		this->selector->OnClick(pt, widget, click_count);
 		this->SetDirty();
-		InvalidateWindowData(WC_INCOME_GRAPH, 0);
-		InvalidateWindowData(WC_OPERATING_PROFIT, 0);
-		InvalidateWindowData(WC_DELIVERED_CARGO, 0);
-		InvalidateWindowData(WC_PERFORMANCE_HISTORY, 0);
-		InvalidateWindowData(WC_COMPANY_VALUE, 0);
 	}
 
-	/**
-	 * Some data on this window has become invalid.
-	 * @param data Information about the changed data.
-	 * @param gui_scope Whether the call is done from GUI scope. You may not do everything when not in GUI scope. See #InvalidateWindowData() for details.
-	 */
-	void OnInvalidateData([[maybe_unused]] int data = 0, [[maybe_unused]] bool gui_scope = true) override
+	void OnInvalidateData(int data = 0, bool gui_scope = true) override
 	{
-		if (!gui_scope) return;
-		if (Company::IsValidID(data)) return;
+		this->selector->OnInvalidateData(data, gui_scope);
+
+		/* Check if this selector window is no longer needed and close it. */
+		if (this->window_number == 0) {
+			/* Close if no company graph windows exist. */
+			if (!FindWindowByClass(WC_INCOME_GRAPH)
```

Comment:
We don't implicitly consider `nullptr` false, you need to check `FindWindowByClass(WC_INCOME_GRAPH) == nullptr`.